### PR TITLE
detectors: Rewrite unitialized state variables with IR.

### PR DESCRIPTION
**Pull request by cryptomental**
Refs: https://github.com/trailofbits/slither/issues/30

### CODE REVIEW

**Comment by montyly** [comment]
Thanks for your contribution!

After some thinking, I think we can go further in the precision of this detector:
- We can remove unused variables by checking if the variable is read somewhere (`operation.read` returns the list of variables read). Unused variables are already detected by `unused-state`
- In its current form, the detector will not track index and mapping.

For example in:
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
both  `balances` and `balancesInitialized` are detected as unitialized. The issue is that index and mapping use references:
```
		Expression: balancesInitialized[msg.sender] = 0
		IRs:
			REF_1(uint256) -> balancesInitialized[msg.sender]
			REF_1 := 0
```

A solution (that I am using in the taint analyses) is to keep track of the reference in a dictionary while iterating over the IR BB:
https://github.com/trailofbits/slither/blob/836f3bab337d12e6e7a92294af92a2a90e398d5c/slither/analyses/taint/common.py#L4-L7

Then when the `lvalue` is a `ReferenceVariable`, we can track back the original variable used, and add this variable to the list of variable written if it's a state variable.

I realize that the handling of rerefences should not be done at the detectors level. Maybe I should add an attribute to ReferenceVariable such as `origin` or `points_to`, or something like that?

- Last case of FP that I can think of, is the case of variable init in a call:
```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test{
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    } 

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

If the variable is used as `storage` parameter, it can be init. Either we need to analyze the call to see if it initialize the variable, or just assume that any call with a storage variable will write to it (if not, it could be another detector). 
On a function call, you can access to the list of parameters with `function.parameters`. The parameters are variables, and have a `location` attribute.

Tldr: If the variable is the parameter of an `LibraryCall`, `InternalCall`, `InternalDynamicCall`, and it declared as `storage`, it should not be detected as unitialized.

Does that make sense?

**Comment by montyly** [comment]
Thanks for your contribution!

After some thinking, I think we can go further in the precision of this detector:
- We can remove unused variables by checking if the variable is read somewhere (`operation.read` returns the list of variables read). Unused variables are already detected by `unused-state`
- In its current form, the detector will not track index and mapping.

For example in:
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
both  `balances` and `balancesInitialized` are detected as unitialized. The issue is that index and mapping use references:
```
		Expression: balancesInitialized[msg.sender] = 0
		IRs:
			REF_1(uint256) -> balancesInitialized[msg.sender]
			REF_1 := 0
```

A solution (that I am using in the taint analyses) is to keep track of the reference in a dictionary while iterating over the IR BB:
https://github.com/trailofbits/slither/blob/836f3bab337d12e6e7a92294af92a2a90e398d5c/slither/analyses/taint/common.py#L4-L7

Then when the `lvalue` is a `ReferenceVariable`, we can track back the original variable used, and add this variable to the list of variable written if it's a state variable.

I realize that the handling of references should not be done at the detectors level. Maybe I should add an attribute to ReferenceVariable such as `origin` or `points_to`, or something like that?

- Last case of FP that I can think of, is the case of variable init in a call:
```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test{
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    } 

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

If the variable is used as `storage` parameter, it can be init. Either we need to analyze the call to see if it initialize the variable, or just assume that any call with a storage variable will write to it (if not, it could be another detector). 
On a function call, you can access to the list of parameters with `function.parameters`. The parameters are variables and have a `location` attribute.

Tldr: If the variable is the parameter of a `LibraryCall`, `InternalCall`, `InternalDynamicCall`, and it declared as `storage`, it should not be detected as unitialized.

Does that make sense?

**Comment by montyly** [comment]
Thanks for your contribution!

After some thinking, I think we can go further in the precision of this detector:
- We can remove unused variables by checking if the variable is read somewhere (`operation.read` returns the list of variables read). Unused variables are already detected by `unused-state`
- In its current form, the detector will not track index and mapping.

For example in:
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
both  `balances` and `balancesInitialized` are detected as unitialized. The issue is that index and mapping use references:
```
		Expression: balancesInitialized[msg.sender] = 0
		IRs:
			REF_1(uint256) -> balancesInitialized[msg.sender]
			REF_1 := 0
```

A solution (that I am using in the taint analyses) is to keep track of the reference in a dictionary while iterating over the IR BB:
https://github.com/trailofbits/slither/blob/836f3bab337d12e6e7a92294af92a2a90e398d5c/slither/analyses/taint/common.py#L4-L7

Then when the `lvalue` is a `ReferenceVariable`, we can track back the original variable used, and add this variable to the list of variable written if it's a state variable.

I realize that the handling of references should not be done at the detectors level. Maybe I should add an attribute to ReferenceVariable such as `origin` or `points_to`, or something like that?

- Last case of FP that I can think of, is the case of variable init in a call:
```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test{
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    } 

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

If the variable is used as `storage` parameter, it can be init. Either we need to analyze the call to see if it initialize the variable, or just assume that any call with a storage variable will write to it (if not, it could be another detector). 
On a function call, you can access to the list of parameters with `function.parameters`. The parameters are variables and have a `location` attribute.

Tldr: If the variable is the parameter of a `LibraryCall`, `InternalCall`, `InternalDynamicCall`, and it declared as `storage`, it should not be detected as unitialized.

Does that make sense?

**Comment by montyly** [comment]
Thanks for your contribution!

After some thinking, I think we can go further in the precision of this detector:
- We can remove unused variables by checking if the variable is read somewhere (`operation.read` returns the list of variables read). Unused variables are already detected by `unused-state`
- In its current form, the detector will not track index and mapping.

For example in:
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
both  `balances` and `balancesInitialized` are detected as unitialized. The issue is that index and mapping use references:
```
		Expression: balancesInitialized[msg.sender] = 0
		IRs:
			REF_1(uint256) -> balancesInitialized[msg.sender]
			REF_1 := 0
```

A solution (that I am using in the taint analyses) is to keep track of the reference in a dictionary while iterating over the IR BB:
https://github.com/trailofbits/slither/blob/836f3bab337d12e6e7a92294af92a2a90e398d5c/slither/analyses/taint/common.py#L4-L7

Then when the `lvalue` is a `ReferenceVariable`, we can track back the original variable used, and add this variable to the list of variable written if it's a state variable.

I realize that the handling of references should not be done at the detectors level. Maybe I should add an attribute to ReferenceVariable such as `origin` or `points_to`, or something like that?

- Last case of FP that I can think of, is the case of variable init in a call:
```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test{
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    } 

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

If the variable is used as `storage` parameter, it can be init. Either we need to analyze the call to see if it initialize the variable, or just assume that any call with a storage variable will write to it (if not, it could be another detector). 
On a function call, you can access to the list of parameters with `function.parameters`. The parameters are variables and have a `location` attribute.

Tldr: If the variable is the parameter of a `LibraryCall`, `InternalCall`, `InternalDynamicCall`, and it declared as `storage`, it should not be detected as unitialized.

Does that make sense?

**Comment by montyly** [comment]
Thanks for your contribution!

After some thinking, I think we can go further in the precision of this detector:
- We can remove unused variables by checking if the variable is read somewhere (`operation.read` returns the list of variables read). Unused variables are already detected by `unused-state`
- In its current form, the detector will not track index and mapping.

For example in:
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
both  `balances` and `balancesInitialized` are detected as unitialized. The issue is that index and mapping use references:
```
		Expression: balancesInitialized[msg.sender] = 0
		IRs:
			REF_1(uint256) -> balancesInitialized[msg.sender]
			REF_1 := 0
```

A solution (that I am using in the taint analyses) is to keep track of the reference in a dictionary while iterating over the IR BB:
https://github.com/trailofbits/slither/blob/836f3bab337d12e6e7a92294af92a2a90e398d5c/slither/analyses/taint/common.py#L4-L7

Then when the `lvalue` is a `ReferenceVariable`, we can track back the original variable used, and add this variable to the list of variable written if it's a state variable.

I realize that the handling of references should not be done at the detectors level. Maybe I should add an attribute to ReferenceVariable such as `origin` or `points_to`, or something like that?

- Last case of FP that I can think of, is the case of variable init in a call:
```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test{
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    } 

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

If the variable is used as `storage` parameter, it can be init. Either we need to analyze the call to see if it initialize the variable, or just assume that any call with a storage variable will write to it (if not, it could be another detector). 
On a function call, you can access to the list of parameters with `function.parameters`. The parameters are variables and have a `location` attribute.

Tldr: If the variable is the parameter of a `LibraryCall`, `InternalCall`, `InternalDynamicCall`, and it declared as `storage`, it should not be detected as unitialized.

Does that make sense?

**Comment by cryptomental** [comment]
Hi @montyly thanks a lot for your hints. Let me digest this and I will correct the detector according to the comments above.

**Comment by montyly** [comment]
I added a `points_to` attribute to `ReferenceVariable` in the `dev` branch, that should help to keep track of all the write

Example:
`pointer.sol`
```
contract Test{
    mapping (address => uint) balances;
    mapping (address => uint) balancesInitialized;


    function init(){
        balancesInitialized[msg.sender] = 0;
    } 

    function use(){
        // random operation to use the mapping
        require(balances[msg.sender] == balancesInitialized[msg.sender]);
    }
}
```
`pointer.py`
```
from slither.slither import Slither
from slither.slithir.operations import OperationWithLValue, Member, Index
from slither.slithir.variables import ReferenceVariable

slither = Slither('pointer.sol')

for contract in slither.contracts:
    for function in contract.functions:
        for node in function.nodes:
            for ir in node.irs:
                # Filter to operation with LValue 
                if isinstance(ir, OperationWithLValue):

                    # Dont consider Member and Index operations
                    # As they create the ReferenceVariable itself
                    # Otherwise, it would consider the creation of the
                    # Referencevariable as a write to the
                    # points_to variable
                    if isinstance(ir, (Member, Index)):
                        continue

                    # If we write to a ReferenceVariable
                    # We loop until finding the original variable
                    if isinstance(ir.lvalue, ReferenceVariable):

                        dest = ir.lvalue
                        while isinstance(dest, ReferenceVariable):
                            dest = dest.points_to
                        print(dest)
```
It prints `balancesInitialized`

Let me know if that helps

**Comment by cryptomental** [comment]
Hi @montyly I did not have time to reply. Thanks a lot! Yesterday I used the other reference approach and it worked also fine:

```
ret = []
        refs = {}
        self.log(f'### Contract {contract.name}')
        for f in contract.all_functions_called + contract.modifiers:
            self.log(f'Function {f.name}')
            for param in f.parameters:
                self.log(f"#### param {param}")
                if param.location == 'storage':
                    self.log(param.location)
            for n in f.nodes:
                for ir in n.irs:
                    if isinstance(ir, OperationWithLValue) and isinstance(ir.lvalue, StateVariable):
                        ret.append(ir.lvalue)
                    elif isinstance(ir, Assignment) and isinstance(ir.lvalue, ReferenceVariable):
                        ret.append(refs[ir.lvalue])
                    elif isinstance(ir, (Index, Member)):
                        refs[ir.lvalue] = ir.variable_left
                    if isinstance(ir, LibraryCall) \
                            or isinstance(ir, InternalCall) \
                            or isinstance(ir, InternalDynamicCall):
                        self.log("## call")
                        self.log(str(f.all_solidity_variables_used_as_args))
                        for v in ir.function.variables_written:
                            self.log(v.name)
                            ret.append(v)
                        for param in f.parameters:
                            self.log("#### param", param)
                            self.log(str(param))
                            if param.location == 'storage':
                                self.log(ir)
                                ret.append(param)
```

The code is still a big WIP as I was stuck in the last case - the one with libraries where the LibraryCall is triggered. 

```
library Lib{

    struct MyStruct{
        uint val;
    }

    function set(MyStruct storage st){
        st.val = 4;
    }

}


contract Test2 {
    using Lib for Lib.MyStruct;

    Lib.MyStruct st;
    Lib.MyStruct stInitiliazed;

    function init(){
        stInitiliazed.set();
    }

    function use(){
        // random operation to use the structure
        require(st.val  == stInitiliazed.val);
    }

}
```

The problem in this scenario is that no parameter is passed and the variable was still reported as uninitialized. I was only able to extract the variable 'st' from the Library call instead of stInitialized.

Plan for today: refactor the refs according to your new feature and I will have a look at the library again.

**Comment by montyly** [comment]
Nice!
So in a call IR you have two things:
- the arguments of the call: `ir.arguments`
- the parameters of the functions `ir.function.parameters`

`ir.arguments` is the list of variables given to the function, `ir.function.parameters` is the parameters definition of the function. 
For a library call, slither convert the caller (the variable) to an arguments

On your example (called `test.sol`):
```
from slither.slither import Slither
from slither.slithir.operations import LibraryCall

slither = Slither('test.sol')

contract = slither.get_contract_from_name('Test2')

init = contract.get_function_from_signature('init()')

for node in init.nodes:
    for ir in node.irs:
        if isinstance(ir, LibraryCall):
            print([(x.name, x.location) for x in ir.function.parameters])
            print([x.name for x in ir.arguments])
```
will print
```
[('st', 'storage')]
['stInitiliazed']
```
I hope that helps?

**Comment by cryptomental** [comment]
Oh yes, *a lot*. Thanks!

**Comment by cryptomental** [comment]
Aha,  @montyly this was the missing piece of the puzzle ;) I also extended the tests. The detector works fine now for all the cases:

```
$ slither tests/uninitialized.sol --detect-uninitialized-state --disable-solc-warnings
INFO:Detectors: Uninitialized state variable in tests/uninitialized.sol, Contract: Uninitialized, Variable: destination, Used in ['transfer']
INFO:Detectors: Uninitialized state variable in tests/uninitialized.sol, Contract: Test, Variable: balances, Used in ['use']
INFO:Detectors: Uninitialized state variable in tests/uninitialized.sol, Contract: Test2, Variable: st, Used in ['use']
```

**Comment by cryptomental** [comment]
I updated only docstrings with the last push.

File: slither/detectors/variables/uninitialized_state_variables.py [hunk 1]
```diff
@@ -24,29 +28,36 @@ class UninitializedStateVarsDetection(AbstractDetector):
...
+    def written_variables(contract):
+        ret = []
+        for f in contract.all_functions_called + contract.modifiers:
+            for n in f.nodes:
+                for ir in n.irs:
+                    if isinstance(ir, (Index, Member)):
+                        continue  # Don't consider Member and Index operations -> ReferenceVariable
+                    elif isinstance(ir, OperationWithLValue) and isinstance(ir.lvalue, StateVariable):
+                        ret.append(ir.lvalue)
+                    elif isinstance(ir, Assignment) and isinstance(ir.lvalue, ReferenceVariable):
```
Comment (montyly) [inline slither/detectors/variables/uninitialized_state_variables.py]: So the way the references are tracked will not catch all the cases (Reference can be the result of any operation, not only `Assignment`).

I simplified the API: https://github.com/trailofbits/slither/pull/58. From the `dev-refactor-nodes-info` branch,  you can directly uses  `node.state_variables_written` (it recovers the info from the IR directly). It should simplify your function

Though `node.state_variables_written` does not consider the write through a library, so you still need to iterate over the IR and check for `storage` variable, which your function already do.

So you can remove the three first `if` branch here, replace by `node.state_variables_written`, and only keep if branch of the line 46

Comment (cryptomental) [reply slither/detectors/variables/uninitialized_state_variables.py]: @montyly this definitely is a good improvement. I will wait till your PR lands in dev and will afterwards rebase mine with the changes needed.

**Comment by montyly** [comment]
https://github.com/trailofbits/slither/pull/58 is merged in `dev`, as discussed that should simplify this PR

**Comment by cryptomental** [comment]
@montyly new API works like a charm ;) I updated the PR, detector uses n.state_variables_written now.
